File: apex_dashboard/apex_dashboard/page/pl_performance_dashboard/pl_performance_dashboard.py
from __future__ import annotations

from typing import Dict, List, Optional, Sequence

import frappe
from frappe import _
from frappe.utils import get_first_day, getdate, nowdate

from apex_dashboard.dashboard import utils as dashboard_utils


@frappe.whitelist()
def get_context(context):
	context.no_cache = 1
	context.page_title = _("P&L Performance Dashboard")
	return context


@frappe.whitelist()
def get_dashboard_data(
	company: Optional[str] = None,
	posting_date: Optional[str] = None,
	from_date: Optional[str] = None,
	to_date: Optional[str] = None,
):
	try:
		posting_date = posting_date or nowdate()
		posting_date = str(getdate(posting_date))

		from_date = from_date or str(get_first_day(posting_date))
		to_date = to_date or posting_date

		data = build_pl_snapshot(company=company, from_date=from_date, to_date=to_date)

		return {
			"success": True,
			"filters": {
				"company": company,
				"posting_date": posting_date,
				"from_date": from_date,
				"to_date": to_date,
			},
			"data": data,
		}
	except Exception as exc:
		frappe.log_error(frappe.get_traceback(), "P&L Performance Dashboard - get_dashboard_data")
		return {
			"success": False,
			"error": _(str(exc)),
		}


def build_pl_snapshot(company: Optional[str], from_date: str, to_date: str) -> Dict[str, object]:
	mapping = dashboard_utils.get_accounts_for_section("pl_performance_dashboard")
	company_currency = dashboard_utils.get_company_currency(company)

	direct_income = _get_period_balances(mapping.get("direct_income", []), company, from_date, to_date)
	indirect_income = _get_period_balances(mapping.get("indirect_income", []), company, from_date, to_date)
	direct_expenses = _get_period_balances(mapping.get("direct_expenses", []), company, from_date, to_date)
	indirect_expenses = _get_period_balances(mapping.get("indirect_expenses", []), company, from_date, to_date)

	direct_income_total = sum(entry["amount"] for entry in direct_income)
	indirect_income_total = sum(entry["amount"] for entry in indirect_income)
	direct_expense_total = sum(entry["amount"] for entry in direct_expenses)
	indirect_expense_total = sum(entry["amount"] for entry in indirect_expenses)

	net_profit = (direct_income_total + indirect_income_total + direct_expense_total + indirect_expense_total)

	kpis = [
		{
			"key": "direct_income",
			"label": _("الدخل المباشر"),
			"totals": _format_totals(direct_income_total, company_currency),
			"indicator": "positive",
		},
		{
			"key": "indirect_income",
			"label": _("الدخل غير المباشر"),
			"totals": _format_totals(indirect_income_total, company_currency),
			"indicator": "info",
		},
		{
			"key": "direct_expenses",
			"label": _("المصروفات المباشرة"),
			"totals": _format_totals(direct_expense_total, company_currency),
			"indicator": "danger" if direct_expense_total < 0 else "info",
		},
		{
			"key": "indirect_expenses",
			"label": _("المصروفات غير المباشرة"),
			"totals": _format_totals(indirect_expense_total, company_currency),
			"indicator": "danger" if indirect_expense_total < 0 else "info",
		},
		{
			"key": "net_profit",
			"label": _("صافي الربح"),
			"totals": _format_totals(net_profit, company_currency),
			"indicator": "positive" if net_profit >= 0 else "danger",
		},
	]

	return {
		"kpis": kpis,
		"sections": [
			_format_section(_("الدخل المباشر"), direct_income, company_currency),
			_format_section(_("الدخل غير المباشر"), indirect_income, company_currency),
			_format_section(_("المصروفات المباشرة"), direct_expenses, company_currency),
			_format_section(_("المصروفات غير المباشرة"), indirect_expenses, company_currency),
		],
		"chart": _build_chart_data(
			direct_income_total,
			indirect_income_total,
			abs(direct_expense_total),
			abs(indirect_expense_total),
			net_profit,
			company_currency,
		),
	}


def _get_period_balances(
	accounts: Sequence[str],
	company: Optional[str],
	from_date: str,
	to_date: str,
) -> List[Dict[str, object]]:
	if not accounts:
		return []

	conditions = [
		"gl.account in %(accounts)s",
		"gl.docstatus = 1",
		"gl.posting_date BETWEEN %(from_date)s AND %(to_date)s",
	]
	values = {
		"accounts": tuple(accounts),
		"from_date": from_date,
		"to_date": to_date,
	}

	if company:
		conditions.append("gl.company = %(company)s")
		values["company"] = company

	query = f"""
		SELECT
			gl.account,
			SUM(gl.credit - gl.debit) AS amount
		FROM `tabGL Entry` gl
		WHERE {' AND '.join(conditions)}
		GROUP BY gl.account
	"""

	rows = frappe.db.sql(query, values=values, as_dict=True)

	results: List[Dict[str, object]] = []
	for row in rows:
		account = row.account
		amount = row.amount or 0
		currency = frappe.get_cached_value("Account", account, "account_currency") or ""
		results.append(
			{
				"account": account,
				"amount": amount,
				"currency": currency,
			}
		)

	return results


def _format_totals(amount: float, currency: str) -> Dict[str, object]:
	return {
		"by_currency": {currency: amount},
		"base": amount,
	}


def _format_section(label: str, entries: List[Dict[str, object]], company_currency: str) -> Dict[str, object]:
	totals = sum(entry["amount"] for entry in entries)
	return {
		"label": label,
		"totals": _format_totals(totals, company_currency),
		"balances": [
			{
				"account": entry["account"],
				"balance": entry["amount"],
				"currency": company_currency,
				"base_balance": entry["amount"],
			}
			for entry in entries
		],
	}


def _build_chart_data(
	direct_income: float,
	indirect_income: float,
	direct_expenses: float,
	indirect_expenses: float,
	net_profit: float,
	currency: str,
) -> Dict[str, object]:
	return {
		"currency": currency,
		"series": [
			{"name": _("الدخل المباشر"), "value": direct_income},
			{"name": _("الدخل غير المباشر"), "value": indirect_income},
			{"name": _("المصروفات المباشرة"), "value": -direct_expenses},
			{"name": _("المصروفات غير المباشرة"), "value": -indirect_expenses},
			{"name": _("صافي الربح"), "value": net_profit},
		],
	}

File: apex_dashboard/apex_dashboard/page/pl_performance_dashboard/pl_performance_dashboard.js
frappe.pages["pl-performance-dashboard"].on_page_load = function (wrapper) {
	frappe.require(
		[
			"assets/apex_dashboard/js/dashboard_common.js",
			"assets/apex_dashboard/css/dashboard_common.css",
		],
		() => init_pl_dashboard(wrapper)
	);
};

function init_pl_dashboard(wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("P&L Performance Dashboard"),
		single_column: true,
	});

	const state = {
		filters: {
			company: frappe.defaults.get_user_default("company"),
			posting_date: frappe.datetime.get_today(),
			from_date: frappe.datetime.month_start(),
			to_date: frappe.datetime.get_today(),
		},
		data: null,
		chart: null,
	};

	const layout = build_pl_layout(page);
	setup_pl_filters(page, state, refresh);

	function refresh() {
		page.set_indicator(__("جارِ التحميل..."), "blue");
		frappe
			.call("apex_dashboard.apex_dashboard.page.pl_performance_dashboard.pl_performance_dashboard.get_dashboard_data", {
				company: state.filters.company,
				posting_date: state.filters.posting_date,
				from_date: state.filters.from_date,
				to_date: state.filters.to_date,
			})
			.then((response) => {
				const payload = response.message || {};
				if (!payload.success) {
					throw new Error(payload.error || __("فشل تحميل بيانات الأداء الربحي."));
				}
				state.data = payload.data || {};
				state.filters = Object.assign({}, state.filters, payload.filters || {});
				render_pl_dashboard(layout, state);
				page.set_indicator(__("جاهز"), "green");
			})
			.catch((error) => {
				console.error(error);
				page.set_indicator(__("خطأ"), "red");
				frappe.msgprint({
					title: __("خطأ"),
					message: error.message || __("تعذر تحميل لوحة الأداء الربحي."),
					indicator: "red",
				});
			});
	}

	refresh();
}

function build_pl_layout(page) {
	const $container = $(page.body);
	$container.empty();

	const $wrapper = $(`
		<div class="gt-dashboard-wrapper">
			<section class="gt-dashboard-section">
				<div class="gt-dashboard-cards" data-role="kpis"></div>
			</section>
			<section class="gt-dashboard-section" data-role="chart">
				<div class="gt-dashboard-subcard">
					<h5>${__("مقارنة الدخل والمصروفات")}</h5>
					<div id="pl-performance-chart" style="min-height: 260px;"></div>
				</div>
			</section>
			<section class="gt-dashboard-section" data-role="sections"></section>
		</div>
	`);

	$container.append($wrapper);

	return {
		kpis: $wrapper.find('[data-role="kpis"]')[0],
		sections: $wrapper.find('[data-role="sections"]')[0],
		chart: $wrapper.find("#pl-performance-chart")[0],
	};
}

function setup_pl_filters(page, state, refresh) {
	const company_field = page.add_field({
		fieldtype: "Link",
		fieldname: "company",
		label: __("شركة"),
		options: "Company",
		default: state.filters.company,
		change() {
			state.filters.company = company_field.get_value();
			refresh();
		},
	});

	const from_field = page.add_field({
		fieldtype: "Date",
		fieldname: "from_date",
		label: __("من تاريخ"),
		default: state.filters.from_date,
		change() {
			state.filters.from_date = from_field.get_value();
			refresh();
		},
	});

	const to_field = page.add_field({
		fieldtype: "Date",
		fieldname: "to_date",
		label: __("إلى تاريخ"),
		default: state.filters.to_date,
		change() {
			state.filters.to_date = to_field.get_value();
			refresh();
		},
	});

	page.add_action_icon("refresh", () => refresh());
}

function render_pl_dashboard(layout, state) {
	render_pl_kpis(layout.kpis, state.data.kpis || []);
	render_pl_sections(layout.sections, state.data.sections || []);
	render_pl_chart(layout.chart, state);
}

function render_pl_kpis(container, kpis) {
	if (!container) return;
	const cards = kpis.map((kpi) => {
		const totals = kpi.totals || {};
		const baseValue = totals.base || 0;
		const byCurrency = totals.by_currency || {};
		const subtitle = Object.keys(byCurrency)
			.map((currency) => DashboardCommon.formatCurrency(byCurrency[currency], currency))
			.join(" • ");

		return {
			title: kpi.label || "",
			value: DashboardCommon.formatCurrency(
				baseValue,
				frappe.defaults.get_default("currency") || frappe.sys_defaults.currency || "EGP"
			),
			subtitle,
			indicator: kpi.indicator || "info",
		};
	});

	DashboardCommon.renderCards(container, cards);
}

function render_pl_sections(container, sections) {
	if (!container) return;
	container.innerHTML = "";

	if (!sections.length) {
		container.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد بيانات مفصلة")}</div>`;
		return;
	}

	sections.forEach((section) => {
		const card = document.createElement("div");
		card.className = "gt-dashboard-subcard";

		const header = document.createElement("div");
		header.className = "gt-dashboard-section__header";
		const title = document.createElement("h4");
		title.className = "gt-dashboard-section__title";
		title.innerText = section.label || "";
		header.appendChild(title);
		header.appendChild(DashboardCommon.buildTotalsRow(section.totals || {}));
		card.appendChild(header);

		const tableContainer = document.createElement("div");
		card.appendChild(tableContainer);
		DashboardCommon.renderGroupTable(tableContainer, section.balances || []);

		container.appendChild(card);
	});
}

function render_pl_chart(container, state) {
	if (!container || !state.data) return;

	const chartData = state.data.chart || {};
	const labels = (chartData.series || []).map((entry) => entry.name || "");
	const values = (chartData.series || []).map((entry) => entry.value || 0);

	if (!labels.length) {
		container.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد بيانات للمخطط")}</div>`;
		return;
	}

	if (state.chart) {
		state.chart.destroy();
		state.chart = null;
	}

	state.chart = new frappe.Chart(container, {
		data: {
			labels,
			datasets: [
				{
					values,
				},
			],
		},
		type: "bar",
		height: 260,
		colors: ["#2563eb"],
		barOptions: {
			stacked: false,
		},
		axisOptions: {
			xAxisMode: "tick",
			yAxisMode: "tick",
			xIsSeries: true,
		},
		tooltipOptions: {
			formatTooltipY: (value) =>
				DashboardCommon.formatCurrency(value, chartData.currency || frappe.sys_defaults.currency || "EGP"),
		},
	});
}

